collector/partitions.go

```diff
@@ -0,0 +1,264 @@
+	Mount     MountInfo  `json:"mount,omitempty"`
+	UUID      string     `json:"uuid,omitempty"`
+	Size      uint64     `json:"size,omitempty"`
+	SubDev    []BlockDev `json:"subdev,omitempty"`
+	Holder    []string   `json:"holder_of,omitempty"`
+	Slave     []string   `json:"slave_of,omitempty"`
+}
+
+// MountInfo is the metadata of a mounted device
+type MountInfo struct {
+	MountPoint string `json:"mount_point,omitempty"`
+	FSType     string `json:"filesystem,omitempty"`
+	Options    string `json:"mount_options,omitempty"`
+}
+
+const sysClassBlock = "/sys/block"
+
+func GetPartitionStats() []BlockDev {
+	partStats := make([]BlockDev, 0)
+	if dirSysBlk, err := os.Lstat(sysClassBlock); err == nil &&
+		dirSysBlk.IsDir() {
+		fi, err := os.Open(sysClassBlock)
+		if err != nil {
+			printErr(err)
+		}
+		blockDevs, err := fi.Readdir(0)
+		if err != nil {
+			printErr(err)
+		}
+		for _, blk := range blockDevs {
+			var blkDev BlockDev
+			if blkDev.getBlockDevice(blk, nil) {
+				partStats = append(partStats, blkDev)
+			}
+		}
+		matchUUIDs(partStats, checkUUIDs())
+		matchMounts(partStats, checkMounts())
+	}
+	return partStats
+}
+
+func (blkDev *BlockDev) getBlockDevice(blk os.FileInfo, parent os.FileInfo) bool {
+	var fullpath string
+	var dev string
+	var err error
+	if parent != nil {
+		fullpath = path.Join(sysClassBlock, parent.Name(), blk.Name())
+		dev = fullpath
+	} else {
+		fullpath = path.Join(sysClassBlock, blk.Name())
+		dev, err = os.Readlink(fullpath)
+		if err != nil {
+			printErr(err)
+		}
+	}
+
+	if strings.HasPrefix(dev, "../devices/virtual/") &&
+		(strings.Contains(dev, "ram") ||
+			strings.Contains(dev, "loop")) {
+		return false
+	}
+
+	// open the dir
+	var fi *os.File
+	if parent != nil {
+		fi, err = os.Open(dev)
+		if err != nil {
+			printErr(err)
+		}
+	} else {
+		fi, err = os.Open(path.Join(sysClassBlock, dev))
+		if err != nil {
+			printErr(err)
+		}
+	}
+	subFiles, err := fi.Readdir(0)
+	if err != nil {
+		printErr(err)
+		return false
+	}
+
+	// check for sub devices
+	for _, subFile := range subFiles {
+		// check if this is a partition
+		if subFile.Name() == "partition" {
+			blkDev.Partition = true
+		}
+
+		// populate subdev
+		if strings.HasPrefix(subFile.Name(), blk.Name()) {
+			var subBlk BlockDev
+			subBlk.getBlockDevice(subFile, blk)
+			blkDev.SubDev = append(blkDev.SubDev, subBlk)
+		}
+	}
+
+	blkDev.Name = blk.Name()
+	blkSize, err := strconv.Atoi(si.SlurpFile(path.Join(fullpath, "size")))
+	if err == nil {
+		blkDev.Size = uint64(blkSize)
+	}
+
+	slaves, holders := listDeps(blk.Name())
+	if len(slaves) > 0 {
+		for _, slave := range slaves {
+			blkDev.Slave = append(blkDev.Slave, slave.Name())
+		}
+	}
+	if len(holders) > 0 {
+		for _, holder := range holders {
+			blkDev.Holder = append(blkDev.Holder, holder.Name())
+		}
+	}
+
+	return true
+}
+
+func listDeps(blk string) ([]os.FileInfo, []os.FileInfo) {
+	fiSlaves, err := os.Open(path.Join(sysClassBlock, blk, "slaves"))
```

Comment:
The slaves and holders path of SubDev is wrong. It will be /sys/block/sdaX/holders, in fact it's /sys/block/sda/sdaX/holders.

Reply:
This is tried for all `sda` and `sdaX` blocks, if there's no holders folder or it's empty, this section is ignored. The same applies to slaves folder.